storage/rocksdb/ha_rocksdb.cc

```diff
@@ -15634,6 +15672,530 @@ void Rdb_compaction_stats::record_end(rocksdb::CompactionJobInfo info) {
   m_history.emplace_back(std::move(record));
 }
 
+/****************************************************************************
+ * Multi-Range-Read implementation based on RocksDB's MultiGet() call
+ ***************************************************************************/
+
+/*
+  Check if MultiGet-MRR can be used to scan given list of ranges.
+
+  @param  seq            List of ranges to scan
+  @param  bufsz     OUT  How much buffer space will be required
+  @param  flags   INOUT  Properties of the scan to be done
+
+  @return
+     HA_POS_ERROR - The scan cannot be done at all
+     Other value  - Number of expected output rows
+*/
+ha_rows ha_rocksdb::multi_range_read_info_const(uint keyno, RANGE_SEQ_IF *seq,
+                                                void *seq_init_param,
+                                                uint n_ranges, uint *bufsz,
+                                                uint *flags,
+                                                Cost_estimate *cost) {
+  ha_rows res;
+  THD *thd = table->in_use;
+
+  // We allow MultiGet-MRR only with these settings:
+  //   optimizer_switch='mrr=on,mrr_cost_based=off'
+  // mrr_cost_based is not supported
+  bool mrr_enabled =
+      thd->optimizer_switch_flag(OPTIMIZER_SWITCH_MRR) &&
+      !thd->optimizer_switch_flag(OPTIMIZER_SWITCH_MRR_COST_BASED);
+
+  res = handler::multi_range_read_info_const(keyno, seq, seq_init_param,
+                                             n_ranges, bufsz, flags, cost);
+
+  if (res == HA_POS_ERROR) {
+    return res; // Not possible to do the scan
+  }
+
+  // Use the default MRR implementation if @@optimizer_switch value tells us
+  // to, or if the query needs to do a locking read.
+  if (!mrr_enabled || m_lock_rows != RDB_LOCK_NONE) {
+    return res;
+  }
+
+  if (keyno == table->s->primary_key) {
+    // We need all ranges to be single-point lookups using full PK values.
+    // (Range scans, like "pk BETWEEN 10 and 20" or restrictions on PK prefix
+    //  cannot be used)
+    bool all_eq_ranges = true;
+    KEY_MULTI_RANGE range{};
+    range_seq_t seq_it;
+    seq_it = seq->init(seq_init_param, n_ranges, *flags);
+    while (!seq->next(seq_it, &range)) {
+      if (!(range.range_flag & UNIQUE_RANGE)) {
+        all_eq_ranges = false;
+        break;
+      }
+      if (table->in_use->killed) {
+        return HA_POS_ERROR;
+      }
+    }
+
+    if (all_eq_ranges) {
+      // Indicate that we will use MultiGet MRR
+      *flags &= ~HA_MRR_USE_DEFAULT_IMPL;
+      *flags |= HA_MRR_SUPPORT_SORTED;
+      *bufsz = mrr_get_length_per_rec() * res * 1.1 + 1;
+    }
+  } else {
+    // For scans on secondary keys, we use MultiGet when we read the PK values.
+    // We only need PK values when the scan is non-index-only.
+    if (!(*flags & HA_MRR_INDEX_ONLY)) {
```

Comment:
Avoid implicit conversions
```suggestion
    if ((*flags & HA_MRR_INDEX_ONLY) == 0) {
```